hoodie-client/src/main/java/com/uber/hoodie/index/bloom/IndexLookUpTree.java

```diff
@@ -0,0 +1,134 @@
+/*
+ *  Copyright (c) 2018 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *
+ */
+
+package com.uber.hoodie.index.bloom;
+
+import java.io.Serializable;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Look up tree implemented as interval trees to search for any given key in N log N time complexity.
+ */
+class IndexLookUpTree implements Serializable {
+
+  private IndexFileRangeNode root;
+
+  /**
+   * @return the root of the tree. Could be {@code null}
+   */
+  public IndexFileRangeNode getRoot() {
+    return root;
+  }
+
+  /**
+   * Inserts a new {@link IndexFileRangeNode} to this look up tree.
+   *
+   * @param newNode the new {@link IndexFileRangeNode} to be inserted
+   */
+  void insert(IndexFileRangeNode newNode) {
+    root = insert(getRoot(), newNode);
+  }
+
+  /**
+   * Inserts a new {@link IndexFileRangeNode} to this look up tree.
+   *
+   * @param root refers to the current root of the look up tree
+   * @param newNode newNode the new {@link IndexFileRangeNode} to be inserted
+   */
+  private IndexFileRangeNode insert(IndexFileRangeNode root, IndexFileRangeNode newNode) {
+    if (root == null) {
+      root = newNode;
+      return root;
+    }
+
+    if (root.compareTo(newNode) == 0) {
```

Comment:
Can you add comments to the following lines of code to explain what's happening here ?